### ISSUE TITLE
[Discussion] Maintain patched version of emscripten in a separate repository for out-of-tree build?

### ISSUE THREAD

**Issue by ryanking13** [pyodide/pyodide#3315]
We often apply patches to emscripten to fix bugs. Since emscripten team makes releases very often, and most of the patches are accepted upstream, these patches are mostly quite temporary. However, after we start to follow the Emscripten update proposal (https://github.com/pyodide/pyodide/issues/2951), we will probably have to keep those patches for a few Pyodide releases even if it is accepted upstream.

We've experienced that some of these patches are critical when building packages (e.g. https://github.com/pyodide/pyodide/pull/2950), so people who build their own package with our out-of-tree build API would sometimes need to use our patched version of emscripten.

Therefore, I think it would be nice if we maintain a patched version of emscripten in a separate repository of our organization: `pyodide/emscripten` or maybe `pyodide/pyodide-emscripten`.

While working on #3309, I found that `emsdk` provides a useful flag `--override-repository` which allows us to easily install tools like emscripten, ccache from a different repository.

### Proposal

1. fork `emscripten` to `pyodide/emscripten`.

- Keep branches (`pyodide-0.21` `pyodide-0.22`, `pyodide-dev`) for building each pyodide version.

2. Adds a new pyodide-build CLI command for installing patched emscripten e.g. `pyodide install-deps emscripten`

- Maybe install it to `$HOME/.pyodide/emsdk/...`?

### Disadvantages

1. Increases our maintenance burden.
2. When applying a new patch, we'll need 2 steps: 1) first apply it to emscripten repository, 2) and then use it in pyodide repository, which will make testing hard.

### Alternatives

Stop patching Emscripten. In that case, we'll have to wait a few months to use the updated Emscripten in Pyodide. But people who build packages out-of-tree can use Emscripten without worrying about compatibility. We can still provide an pyodide-build CLI installing emscripten, but we can install it from the upstream repository.

Related:

- #1977
- #3287

WDYT?

**Comment by hoodmane** [comment]
Well there are some merits to this proposal. I think I'm in favor of it since it's tidier than what we currently do.

But the vast majority patches make no difference for out of tree builds -- we almost always patch JavaScript code or C stdlibs rather than the compiler itself. I've tested a lot of different package builds with dropped patches and the only problems I've run into is when we have to change a port.

**Comment by ryanking13** [comment]
> But the vast majority patches make no difference for out of tree builds -- we almost always patch JavaScript code or C stdlibs rather than the compiler itself. I've tested a lot of different package builds with dropped patches and the only problems I've run into is when we have to change a port.

Right, that's a good point. Probably it is okay if we don't apply patches that affect dynamic linking or a compiler.

**Comment by ryanking13** [comment]
We now include Emscripten patches in the xbuildenv, and utilizing it in [pyodide-recipes](https://github.com/pyodide/pyodide-recipes/blob/main/tools/install_and_patch_emscripten.py) when building emscripten. There is no official/public APIs for that yet, but I think this is okay to be closed as resolved.
